[FLINK-8354] [flink-connectors] Add ability to access and provider Kafka headers

alexeyt820 opened the pull request:
## What is the purpose of the change
This changes allows to access Kafka headers in KeyedDeserializationSchema, and to provide Kafka headers via KeyedSerializationSchema.headers method

## Brief change log
- Add default _headers_ method to KeyedSerializationSchema interface
- Add default deserialize method with _headers_ argument to KeyedDeserializationSchema
- Add Kafka011Fetcher to get headers from Kafka ConsumerRecord and provide them as parameter KeyedDeserializationSchema
- Modify FlinkKafkaProducer011 to request _headers_ from KeyedSerializationSchema and add them to ProducerRecord.

## Verifying this change

This change added tests and can be verified as follows:
  - *Added integration test *Kafka011ITCase.testHeaders* to verify that we can produce headers and consume them

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (yes)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

yanghua commented on flink-connectors/flink-connector-kafka-0.11/src/test/java/org/apache/flink/streaming/connectors/kafka/Kafka011ITCase.java:
```diff
@@ -350,4 +363,201 @@ public boolean isEndOfStream(Long nextElement) {
+		}
+
+		@Override
+		public TestHeadersElement deserialize(byte[] messageKey, byte[] message, String topic, int partition, long offset, Iterable<Map.Entry<String, byte[]>> headers) throws IOException {
```
too long

alexeyt820 replied:
Thank you. Will do

yanghua commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka09Fetcher.java:
```diff
@@ -206,6 +208,9 @@ protected KafkaConsumerCallBridge createCallBridge() {
 
+	protected Iterable<Map.Entry<String, byte[]>> headersOf(ConsumerRecord<byte[], byte[]> record) {
+		return Collections.emptyList();
+	}
```
insert a new line after this looks better

yanghua commented on flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/util/serialization/KeyedDeserializationSchema.java:
```diff
@@ -54,4 +55,20 @@
+	 * Deserializes the byte message.
+	 *
+	 * @param messageKey the key as a byte array (null if no key has been set).
+	 * @param message The message, as a byte array (null if the message was empty or deleted).
```
missed topic parameter description.

alexeyt820 commented:
@yanghua, @aljoscha any other suggestions?

tzulitai commented:
@alexeyt820 I think it is not easily possible to resolve the conflicts between this PR and #6105, which also touches the `KeyedDeserializationSchema` and follows a somewhat different approach.

I'm also not found of wrapping the Kafka record bytes into `Record`; that basically always ties deserialization of the bytes with access to other meta information, and makes it hard to reuse some already existing deserialization formats such as `AvroDeserializationSchema`.

Is it ok if I open a new PR that is based on this one and #6105, and see what I come up with? I might not be able to do that this week, but I can try next week.

cc @FredTing would that also be ok for you?

FredTing commented:
@tzulitai That's fine with me. Just let me know whether I should close PR #6105 now or wait until you submitted the new PR.

alexeyt820 commented:
@tzulitai, using ```Record``` wrapping Kafka ```ConsumerRecord``` allows to add for example timestamp from PR #6105 w/o need to change client code, so it looks like more extensible approach. Not sure how it makes hard to reuse some already existing deserialization formats such as ```AvroDeserializationSchema```, at least not harder then now - ```AvroDeserializationSchema``` will be wrapped via ```KeyedDeserializationSchemaWrapper``` in exactly same way as know. Also ```KeyedDeserializationSchemaWrapper``` calls only ```Record.value()```, so it doesn't ties deserialization of byte with access to other metadata, not in terms of execution path (in logical terms it is always tied because underlying level - Kafka ```ConsumerRecord``` contains key, value and metadata)

aljoscha commented:
For this, I would await the outcome of #6577. If we end up only having one "modern" Kafka connector we might just pass through the `ConsumerRecord`. Otherwise we might have to do some wrapping.

alexeyt820 commented:
Rebased on top of #6703 and implemented KeyedDeserializationSchema.Record for Kafka 2.0 (had to keep KeyedDeserializationSchema.Record because KeyedDeserializationSchema is used by legacy connectors as well

alexeyt820 commented:
@yanghua, @aljoscha is anything else need to be done with this PR ?

allenxwang commented:
This is a very useful feature for us. Any chance this would be merged soon?

stevenzwu commented on flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/util/serialization/KeyedDeserializationSchema.java:
```diff
@@ -32,6 +34,49 @@
+	 * Kafka record to be deserialized.
+	 * Record consists of key,value pair, topic name, partition offset, headers and a timestamp (if available)
+	 */
+	interface Record {
```
I am wondering if we should just use ConsumerRecord from kafka directly. this `KeyedDeserializationSchema` is for kafka consumer only. there isn't much benefit of defining another Flink interface.

just to give a counter point. I am also interested in the `timestamp` fields. now we need to expand this interface.

alexeyt820 replied:
I already answered before: we had to keep KeyedDeserializationSchema.Record because KeyedDeserializationSchema is used by legacy connectors, e.g. Kafka 0.8 doesn't have _ConsumerRecord_ at all
Regarding timestamp - yes, we will need to expand interface, and I planned address it in separate PR on top of this one.

stevenzwu replied:
@alexeyt820 as pointed out by @azagrebin, kafka 0.8 seems to have `ConsumerRecord` interface
https://archive.apache.org/dist/kafka/0.8.2-beta/java-doc/org/apache/kafka/clients/consumer/ConsumerRecord.html

As Flink is moving to the direction of just one modern kafka connector, I am also wondering if it is ok to drop kafka 0.8 connector?

stevenzwu replied:
We can add a new deserialize method to `KeyedDeserializationSchema` interface with a default implementation that just forwards to the other deserialize method (mark as deprecated).

{code}
T deserialize(ConsumerRecord consumerRecord) throws IOException {
    return deserialize(messageKey, message, topic, partition, offset);
}
{code}

alexeyt820 replied:
Done in 2adbacb

azagrebin commented on flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java:
```diff
@@ -641,6 +641,9 @@ public void invoke(KafkaTransactionState transaction, IN next, Context context)
 		} else {
 			record = new ProducerRecord<>(targetTopic, null, timestamp, serializedKey, serializedValue);
 		}
+		for (Map.Entry<String, byte[]> header: schema.headers(next)) {
```
tab before `:`

alexeyt820 replied:
Pushed e0ca3e0

alexeyt820 commented on flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java:
```diff
@@ -251,6 +269,18 @@ public FlinkKafkaConsumerBase(
+	 * @param properties - Kafka configuration properties to be adjusted
+	 * @param offsetCommitMode offset commit mode
+	 */
+	protected static void adjustAutoCommitConfig(Properties properties, OffsetCommitMode offsetCommitMode) {
```
Pushed 3cfda16

alexeyt820 commented on flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java:
```diff
@@ -82,93 +84,110 @@
-		CheckpointListener,
-		ResultTypeQueryable<T>,
-		CheckpointedFunction {
+	CheckpointListener,
```
Yes, sorry. Looks like I intentionally reformat it. Will revert.

alexeyt820 replied:
522589e

alexeyt820 commented on flink-connectors/flink-connector-kafka-0.8/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/SimpleConsumerThread.java:
```diff
@@ -370,8 +370,8 @@ else if (partitionsRemoved) {
 
-							final T value = deserializer.deserialize(keyBytes, valueBytes,
-									currentPartition.getTopic(), currentPartition.getPartition(), offset);
+							final T value = deserializer.deserialize(
```
Thank you for suggestion, I will try

alexeyt820 replied:
Methods of 0.8 Consumer have different signatures compared to modern Kafka - namely they throw base Exception.

1. Change signature of deserialize(ConsumerRecord) to throw Exception instead of IOException. 
2. Wrap and re-throw in every deserialize(ConsumerRecord)  implementation which is 0.8 or for Kafka-base (KeyedDeserializationSchemaWrapper, JSONKeyValueDeserializationSchema, NonContinousOffsetsDeserializationSchema and TypeInformationKeyValueSerializationSchema) 

@azagrebin, @stevenzwu which approach do you prefer?

alexeyt820 replied:
Done in 2adbacb

azagrebin commented on flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/util/serialization/KeyedSerializationSchema.java:
```diff
@@ -55,4 +57,13 @@
+	 * @param element The incoming element to be serialized
+	 * @return collection of headers (maybe empty)
+	 */
+	default Iterable<Map.Entry<String, byte[]>> headers(T element) {
```
I would even suggest to rewrap Kafka Header class into Flink class here, because it should be more extensible if Kafka decides to extend its Header in future. But if it is possible to accept just Kafka `ConsumerRecord` directly for all versions then we do not need any `Record` wrapping at all.

alexeyt820 replied:
Not sure how calling one constructor can be easier than calling another... Map.Entry is more convenient in case headers representation in element is Map (or Guava's Multimap), in that case it just particular implementation would be element.entrySet()

alexeyt820 replied:
I can try to use ProducerRecord by adding factory method to KeyedSerializationSchema
```suggestion
interface KeyedSerializationSchema {
...
	default ProducerRecord<byte[], byte[]> newProducerRecord(T element, 
	byte[] serializedKey, byte[] serializedValue, Integer partition, String topic, Long timestamp) {
		return new ProducerRecord<>(topic, partition, timestamp, serializedKey, serializedValue);
	}
}
```
And then in a producer:
```suggestion
	String targetTopic = schema.getTargetTopic(next);
	if (targetTopic == null) {
		targetTopic = defaultTopicId;
	}
	...
	ProducerRecord<byte[], byte[]> record = schema.newProducerRecord(
	  next, 
	  serializedKey,
	  serializedValue
	  flinkKafkaPartitioner != null ?
		flinkKafkaPartitioner.partition(next, serializedKey, serializedValue, targetTopic, partitions) : null,
	  targetTopic, 
	  timestamp);
	  ...
```
Note that we can't deprecate getTargetTopic, serializeKey and serializedValue, because they are used by partitioner as well.

What do you think?

alexeyt820 replied:
Can't have default method creating `ProducerRecord`, because 0.8 doesn't have timestamp parameter in ProducerRecord constructor

alexeyt820 replied:
@azagrebin, so you suggest to have Flink class to represent Kafka Header instead of Map.Entry?

azagrebin replied:
@alexeyt820 
Ideally we could deprecate the partitioner in the producer constructer as well because ProducerRecord already contains partition which user can assign for the record.

Instead of deprecating the methods in (de)ser schema interfaces we could deprecate them fully and introduce new `Kafka(De)SerializationSchema` interfaces which work with Kafka `Consumer/ProducerRecord` classes.

We also introduce adaptors from older schemas to the newer ones. The producer/consumer constructors, which currently accept older schemas, will use adaptors to create newer schemas. The actual code should also work with newer schemas only.

The serialization schema adaptor could optionally take the `FlinkKafkaPartitioner` to populate the partition in similar way as now it happens directly in producer.

What do you think?

alexeyt820 replied:
Let's say we have 
```
interface KafkaSerializationSchema<T> {
  ProducerRecord toKafka(T element)
```
Adapter for KafkaSerializationSchema will be Kafka version specific, so we will need to have `KafkaSerializationSchemaAdapter08<T>`, `KafkaSerializationSchemaAdapter09<T>`.
Also if we replacing partitioner we need to take into account, that the partitioner had initialization method - open and extract arguments, so it will become something like that
```
interface KafkaSerializationSchema<T> {
  void open(int parallelInstanceId, int parallelInstances);
  ProducerRecord toKafka(T element, int[] partitions);
```
in my opinion it doesn't look clean. 

Alternative, we can keep Partitioner as is, but introduce version specific `ProducerRecord reparation( ProducerRecord source, int partition)` in _FlinkKafkaProduceXY_.
Then FlinkKafkaProducer can do following:
```
  ProducerRecord record = schema.toKafka(value);
  if (flinkKafkaPartitioner != null) {
   int partition = flinkKafkaPartitioner.partition(value, record.key(), record.value(), record.topic(), partitions);
  if(record.partition() == null || partition != record.partition()) {
    record = this.reparation(record,partition);
  }
```
So if we have `flinkKafkaPartitioner` and if it chooses different partition, we will have extra object allocation. Maybe not so bad?
What is your opinion?

azagrebin replied:
Let's say the new serialization schema looks like this:
```
interface KafkaSerializationSchema<T> {
  ProducerRecord serialize(T element, @Nullable Long timestamp, PartitionInfo partitionInfo);
}
class PartitionInfo {
  private final int parallelInstanceId, int parallelInstances, int[] partitions);
  // ....
}
```
This way it should preserve previous Partitioner API capabilities.

Regarding the multiple adaptors, from what I see, the difference is that the older connectors do not support timestamp in `ProducerRecord`. Could we have one adaptor in base module which calls `ProducerRecord` constructor without timestamp if `KafkaSerializationSchema.serialize` is called with `timestamp = null`? Producers, which do not supported timestamp, could call new schema with `timestamp = null`, including base one (although `ProducerRecord` has `timestamp` in base).

Regarding the open method, could adaptor or partitioner still be opened the similar way as it happens now to current partitioner? I agree that the new schema would look weird with the open method.

`PartitionInfo` could be populated in a similar way as `int[] partitions` are cached now in `FlinkKafkaProducer.topicPartitionsMap` in `FlinkKafkaProducer.invoke`, just `parallelInstanceId` and `parallelInstances` are added.

alexeyt820 replied:
So we still need two KafkaSerializationSchema adapters: one for base, 0.8 and 0.9 which created ProducerRecord w/o timestamp and one with timestamp for rest of versions?

Regarding `open`, I guess we can do it, but if someone had non-trivial overridden `open` and want to migrate to new `KafkaSerializationSchema` it will be more challenging, because he will need to move logic of `open`, which was called once into `serialize`, which is called for each record. Likely rare case, and it is feasible to cache results of doing "open" during `serialize` in `KafkaSerializationSchema`. So let's consider it acceptable?

azagrebin replied:
Yes, I would try with 2 adaptors and accept the migration risk. API w/o open method looks simpler. It can be extended later anyways if we discover use cases for it.

alexeyt820 replied:
What about `defaultTopic`? We need it, because `KeyedSerializationSchema.getTargetTopic` may (and I guess usually does) return null. We can add `defaultTopic` as parameter to `KafkaSerializationSchema.serialize` or we can add it to `PartitionInfo`. 

And `int[] partitions` seems should be method `int[] PartitionInfo.partitions(String topicName), because `KeyedSerializationSchema.serialize` need ability to query partitions for arbitrary topic.

alexeyt820 replied:
Also we can provide `defaultTopic` via constructor to an adapter. And for assume similar approach for other `KeyedSerializationSchema` implementations.

azagrebin replied:
True, it should be `PartitionInfo.getPartitions(topic)` and the adaptor has to have the `defaultTopic`.
Maybe, we could pack `topicPartitionsMap` and `getPartitionsByTopic` logic into `PartitionInfo` which would be one object belonging to `FlinkKafkaProducer`.

alexeyt820 replied:
Or may be it should be `ProducerRecord serialize(T element, String defaultTopic, @Nullable Long timestamp, PartitionInfo partitionInfo);`, otherwise every implementing class will need to have `topic` as constructor argument.

azagrebin replied:
I would keep it in the constructor because it will be the same for each `serialize` method. We could consider having overloaded `PartitionInfo.getPartitions()` method but this would also mean some implicit semantics of `PartitionInfo` having `defaultTopic` from `FlinkKafkaProducer`.

alexeyt820 replied:
4db60d2

azagrebin replied:
Thanks, I will have a look. 

Do you also plan to introduce the similar `KafkaDeserialisationSchema` to covert `ConsumerRecord` into object to have symmetrical API?

Also, it would be nice to split this big PR into two PRs with separate consumer and producer refactoring. The consumer part seems to be more clear and popular as a feature, we could try to merge it first.

alexeyt820 replied:
I didn't plan to introduce `KafkaDeserialisationSchema`, current approach with default method is simpler, but if you think that it is important to have symmetrical API, I can do it. Let me know.

I don't like idea of splitting into two PRs, currently added tests produce records with headers and consumes them. I feel splitting will add extra complexity at least from unit test standpoint. And for my purposes I need both - consumer and producer with headers.

azagrebin replied:
The change for `KafkaDeserialisationSchema` might be simple if we do it like this:
```
// new API, that we could now use everywhere in Kafka code instead of KeyedDeserializationSchema
interface KafkaDeserialisationSchema<T> {
  T deserialize(ConsumerRecord<byte[], byte[]> record);
}

// old API, stays almost the same as we already have it in the PR and also supports the new API
@deprecated
interface KeyedDeserializationSchema<T> extends KafkaDeserialisationSchema<T> {
  T deserialize(byte[] messageKey, byte[] message, String topic, int partition, long offset);
  default T deserialize(ConsumerRecord<byte[], byte[]> record) {
    return deserialize(record.key(), record.value(), record.topic(), record.partition(), record.offset());
  }
}
```
Do you think it is quick enough change?

Ok, I see your point with tests regarding the splitting. The thing is that we are interested in merging this PR this week before feature freeze for 1.8 release (planned this Friday). Let's see then how it goes and how much we can manage. Do you have time this week to work on this?

alexeyt820 replied:
It is quick enough. Keep in mind that it also affects all constructors etc, so it is more then 2 files change. And there were conflicting changes in `KafkaConsumerBase` and `KafkaConsumerBaseTest` so had to merge upstream master in PR branch.
Hopefully 7e6753b will be ok

alexeyt820 commented:
Weird build is fine locally using 1.8.0_181-b13, but fails on Travis with 1.8.0_151-b12:

>17:38:32.202 [ERROR] /home/travis/build/apache/flink/flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/JSONKeyValueDeserializationSchemaTest.java:[105,71] cannot infer type arguments for org.apache.kafka.clients.consumer.ConsumerRecord<>

for following lines:
```
		final ConsumerRecord<byte[], byte[]> consumerRecord = new ConsumerRecord<>(
			"topic#1", 3, serializedKey, serializedValue, 4L);

```

Any ideas?

alexeyt820 commented:
The compile problem was due master flink-connector-kafka-base uses Kafka 0.10.2.1, where ConsumerRecord constructor has different signature from Kafka 0.8, which was used flink-connector-kafka-base when I branched PR branch. To be signature independent, changed test to use mock instead for creating ConsumerRecord via constructor.

alexeyt820 commented on flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka011ConsumerRecord.java:
```diff
@@ -0,0 +1,57 @@
+
+	@Override
+	public Iterable<Map.Entry<String, byte[]>> headers() {
+		return Iterables.transform(consumerRecord.headers(), HEADER_TO_MAP_ENTRY_FUNCTION);
```
Resolved (file removed) in 2adbacb

aljoscha commented:
@alexeyt820 Please don't push any more changes right now. I'm having a look at this PR and might shuffle a few things around and rebase it.

alexeyt820 commented:
@aljoscha, sure. Let me know if I can help

aljoscha commented:
I created rebased PRs that have separate commits to be easier to review: #7781 and #7782

aljoscha commented:
Closing because we have the other PRs.

Thanks for pushing this @alexeyt820!
